app/code/community/Bolt/Boltpay/Block/Checkout/Boltpay.php

```diff
@@ -168,20 +163,113 @@ public function getCartDataJs($multipage = true)
 
             if($multipage) {
                 // Resets shipping rate
-                $shippingMethod = $quote->getShippingAddress()->getShippingMethod();
-                $boltHelper->applyShippingRate($quote, null);
+                $shippingMethod = $sessionQuote->getShippingAddress()->getShippingMethod();
+                $boltHelper->applyShippingRate($sessionQuote, null);
             }
 
             // Call Bolt create order API
             try {
-                $orderCreationResponse = $this->createBoltOrder($quote, $multipage);
+                /////////////////////////////////////////////////////////////////////////////////
+                // We create a copy of the quote that is immutable by the customer/frontend
+                // Bolt saves this quote to the database at Magento-side order save time.
+                // This assures that the quote saved to Magento matches what is stored on Bolt
+                // Only shipping, tax and discounts can change, and only if the shipping, tax
+                // and discount calculations change on the Magento server
+                ////////////////////////////////////////////////////////////////////////////////
+
+                /*********************************************************/
+                /* Clean up resources that may have previously been saved
+                /* @var Mage_Sales_Model_Quote[] $expired_quotes */
+                $expired_quotes = Mage::getModel('sales/quote')
+                    ->getCollection()
+                    ->addFieldToFilter('parent_quote_id', $sessionQuote->getId());
+
+                foreach( $expired_quotes as $expired_quote) {
+                    $expired_quote->delete();
+                }
+                /*********************************************************/
+
+                /* @var Mage_Sales_Model_Quote $immutableQuote */
+                $immutableQuote = Mage::getSingleton('sales/quote');
+
+                try {
+                    $immutableQuote->merge($sessionQuote);
+                } catch ( Exception $e ) {
+                    Mage::helper('boltpay/bugsnag')->notifyException($e);
+                }
+
+                if (!$multipage) {
+                    // For the checkout page we want to set the
+                    // billing and shipping, and shipping method at this time.
+                    // For multi-page, we add the addresses during the shipping and tax hook
+                    // and the chosen shipping method at order save time.
+                    $immutableQuote
+                        ->setBillingAddress($sessionQuote->getBillingAddress())
+                        ->setShippingAddress($sessionQuote->getShippingAddress())
+                        ->getShippingAddress()
+                        ->setShippingMethod($sessionQuote->getShippingAddress()->getShippingMethod())
+                        ->save();
+                }
+
+                $already_set_by_merge = array(
```

Comment:
why we need this?

What is the problem of doing like:

```
foreach($sessionQuote->getData() as $key => $value ) {
  $immutableQuote->setData($key, $value);
}
```

Reply:
When running through the array completely, a critical element is being overwritten that causes the order creation to fail as the cart total passed to Bolt create order is set to zero.  All of these data points are set by merge in a working condition and do not need to be reset.

Each item can be checked one by one to see which one or which ones are causing the problem, but, they could all just as easily be ignored and not reset.  I chose the latter option.

The reason for going through the array and setting the values not listed is because other plugins may have added extra fields to the quote table like us, and we want to preserve this.

Reply:
So, it is not the prettiest looking piece of code, but, I can't really justify going through every element in the array that is already set when I know that simply not overwriting them will not produce the error that occurs when they are overwritten.  It may be that overwriting certain values trigger certain events, or other things that causes problems -- and the problems are clearly introduced by the unchecked
`foreach($sessionQuote->getData() as $key => $value ) {
  $immutableQuote->setData($key, $value);
}`

Reply:
how about moving this and foreach below into subfunction named `copyAdditionalData` and add comment ther ethat just coping all data via foreach can cause weird result?